### PR TITLE
Introduce cop_codegen to simplify some code

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

Extracted from https://github.com/tikv/tikv/pull/3898

This PR moves macro into a custom derive, to further simplify the defining process of Rpn functions and batch aggregate functions.

This PR also includes a simple code simplification, changing `xxx.map(|..| Error::Other(box_err!(..)))?` to `box_try!(xxx)`.

- [x] Merge https://github.com/tikv/tikv/pull/4626 first to resolve CI issue

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Existing tests.

### CODE REVIEW

**Comment by breeswish** [comment]
CI failed because CI uses check-bin script from master instead of this PR. Will be fixed later.

File: components/cop_codegen/src/rpn_function.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+#[derive(FromDeriveInput, Debug)]
+#[darling(attributes(rpn_function))]
+pub struct RpnFunctionOpts {
+    ident: syn::Ident,
+    generics: syn::Generics,
+    #[darling(default)]
```
Comment (kennytm) [inline components/cop_codegen/src/rpn_function.rs]: I suggest removing the default, requiring explicit `#[rpn_function(args = 0)]`, making the reader easier to understand.

File: components/cop_codegen/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+use proc_macro::TokenStream;
+use syn::DeriveInput;
+
+#[proc_macro_derive(RpnFunction, attributes(rpn_function))]
+pub fn rpn_function_derive(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+    let r = rpn_function::RpnFunctionOpts::from_derive_input(&input);
+    match r {
+        Err(e) => panic!("{}", e),
+        Ok(r) => TokenStream::from(r.generate_tokens()),
```
[the remark this answers is not in the document]
Comment (breeswish) [reply components/cop_codegen/src/lib.rs]: Yes I would like a simple solution, but unwrap output errors using `Debug` (and the Error structure provided by darling is quite complicated), while `Display` makes error more friendly.

File: components/cop_codegen/src/rpn_function.rs (hunk 1 above)
Comment (breeswish) [reply components/cop_codegen/src/rpn_function.rs]: Nice suggestion!

File: components/cop_codegen/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kennytm) [reply components/cop_codegen/src/lib.rs]: I see. (You could use `r.unwrap_or_else(|e| panic!("{}", e)).generate_tokens().into()` but this is not much better than the `match` then.)

**Pull request merged**